param_passing/proposal.md

```diff
@@ -0,0 +1,194 @@
+
+## Proposed changes
+
+Parameters will be specified by the user in either model or world files through the use of model [composition (or nested models)](http://sdformat.org/tutorials?tut=composition) which will contain `//include` tags and [custom elements](http://sdformat.org/tutorials?tut=custom_elements_attributes_proposal&cat=pose_semantics_docs). Discussed in this section is the proposed design for specifying parameters and the actions available to the user. The intent is to provide a means of manipulating almost all parameters in SDF models through model/world files without requiring modifications to already existing/created models files.
+
+### 1. Add custom element `//include/experimental:params`
+
+In a launch/world file, the user can construct a model using model composition and include components of the model using the `//include` tag. If a user wants to modify parameters of the included model, then the user will use the experimental custom element `//include/experimental:params` that will describe the adjustments to elements. In the custom element `//include/experimental:params`, `experimental` is the namespace prefix and `params` is the custom element. This approach was chosen so that downstream users can choose to ignore the namespaced custom element and its contents. The goal is to implement passing additional arguments in `//include` through an experimental custom element in SDFormat 1.7 / `libsdformat` 10 then after being vetted it will be made official in a future SDFormat / `libsdformat` release where the custom element will be changed to `//include/params`.The included model (not the constructed model in the model/world file) will be referred to as the original model/file. In the model/world file and under `//include`, the elements listed under `//include/experimental:params` will indicate elements from the original model and will specify new values and/or elements to be updated, added, and/or removed.
+
+**Details**
+
+To specify an element from the original file, the tag (e.g., `//link`, `//sensor`, `//visual`) needs to be provided as well as the name attribute but the name is the name of all the parent elements including its own namespace separated by double colon (`::`) syntax (e.g., `name="chassis::camera::cam"` where "chassis" is the name of the `//link`, "camera" is the name of the `//link/sensor`, and "cam" is the name of the `//link/sensor/camera`). This is how the correct element will be identified and will be called the element identifier. Although it is possible to identify the element without requiring the user to specify the tag, enforcing the user to provide the tag allows less ambiguity to the user as well as provides a means of error checking. Then the user will need to specify a corresponding action with that element to dictate the alteration they would like to make using a new attribute `action`. The attribute takes in a string and the available actions at the specified element are:
```

Comment:
The `::` delimiter is slated to be disallowed for `name` attributes (https://github.com/osrf/sdformat/issues/420) in SDFormat 1.8. Using `name` as the path to the element could lead to confusion. Could this be changed to something else? eg. `path`, `element_path`, `ref`, `element_id`, etc.

Reply:
I updated it to `element_id` 84cc9e0cb3e60071a62ce3d0769286de5d2d5464